### ISSUE TITLE
Add a Resource Monitor in Overload Manager that Tracks Memory PSI

### ISSUE THREAD

**Issue by ryansmick** [envoyproxy/envoy#36681]
*Title*: *Add a Resource Monitor in Overload Manager that Tracks Memory PSI*

*Description*:
I’m currently using Envoy as part of Istio, and we are attempting to use Overload Manager to prevent envoy from using too much memory which could result in having the istio-proxy container OOM killed.

The method we’re currently evaluating is to track memory via the fixed_heap monitor, and take appropriate action to slow memory growth based on the pressure reported by that monitor. As istio-proxy runs with multiple processes inside it, it can be difficult to determine the correct value at which to set the max heap size. If we set it too low, we leave resources on the table and prematurely throttle. Worse, if we set it too high, we risk envoy and the container being OOM killed before our throttling kicks in at all.

Linux has a feature called [Pressure Stall Information](https://docs.kernel.org/accounting/psi.html#psi) (or “PSI” for short) that quantifies overcommitted resources (including CPU, memory, and I/O) in terms of how long processes are stalled waiting for those resources. This PSI, as you can see, is also available per cgroup in cgroup v2. Having a resource monitor that tracks this memory pressure would allow us to dynamically understand when we’re approaching a memory threshold for the container rather than setting static thresholds and hoping they’re correct.

I understand adding a new resource monitor, especially one that has dependencies on something as specific as cgroup v2, will likely require discussion, even before design and implementation, so I wanted to open this issue to begin that discussion of whether the community would be open to adding this as a resource monitor in Envoy.

**Comment by tyxia** [comment]
add @nezdolik and @KBaichoo , as domain/code owner

**Comment by nezdolik** [comment]
This sounds like a useful feature (which will not be available for Osx users). What would be the reasonable action for Envoy process/new resource monitor if another process within container/VM is consuming too much memory? When using this PSI stat one cannot deduce which exact process is causing memory stall for other processes, correct? Also, please get familiar with our extension policy: https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md#adding-new-extensions

**Comment by KBaichoo** [comment]
Can you help me understand exactly when a task would end up stalling on memory from the PSI interface? e.g. does this mean swap kicks?  OOM killer kicks in? the task is blocked by page faulting? 

>it can be difficult to determine the correct value at which to set the max heap size. If we set it too low, we leave resources on the table and prematurely throttle. Worse, if we set it too high, we risk envoy and the container being OOM killed before our throttling kicks in at all.

What if you instead created a cgroup based heap monitor that would dynamically adjust to the resources provides to the group -- by reading the value itself and providing some fixed percent of that allocated for envoy to react.

I'm not an expert in memory PSI, a downsides I see is the following: the granularity seems to be avg 10, 60 and 300 seconds which might be too coarse for sharp spikes in resource usage

**Comment by KBaichoo** [comment]
Thank you @ryansmick for all of the resources and pointers (especially the timestamped section!). This is a good idea. I agree that on the surface this looks more user friendly than trying to find the “correct” limits or trying to handle burstiness of memory usage if going above the high threshold doesn’t actually cause any system problems. 

One aspect that I don’t yet have an intuitive understanding of is “how much time stalling due to memory pressure” is too much. 

I look forward to learning more about your progress on this :).

**Comment by ryansmick** [comment]
@KBaichoo Ok, I have some data to share. There's a lot of information, so I'll spread it over a few comments.

I ran some experiments with a AWS EC2 instance running cgroup v2. Each experiment had a different pattern of memory allocation of different types. TLDR of learnings to come at the end of all the comments.

# Test Bench
OS Version: Ubuntu 24.04
CPU Arch: x86
Kernel Version: 6.8.0-1016-aws
Cgroup memory limit (memory.max): 128Mi
Cgroup memory.high: None
Cgroup Swap Space (memory.swap.max): None

## Anonymous - No Swap, No memory.high
Continuously allocate anonymous memory on the heap in 200Ki chunks

![Memory Usage Over Time](https://github.com/user-attachments/assets/5aaee31a-abc5-4c57-8f41-6e87edf41b6c)
![PSI Change Over Time](https://github.com/user-attachments/assets/11631015-0a5d-4e38-96a0-c87933f070dd)

## Clean File - No Swap, No memory.high
Perform buffered read of 128Mi file, and then continuously allocate anonymous memory on the heap in 200Ki chunks

![Memory Usage Over Time](https://github.com/user-attachments/assets/837845c0-fff0-45b2-94cb-412c056ad593)
![PSI Change Over Time](https://github.com/user-attachments/assets/68299841-37f0-4276-b7b5-e54e591d2446)

## Clean Mmaped File - No Swap, No memory.high
Perform buffered read of 128Mi memory-mapped file, and then continuously allocate anonymous memory on the heap in 200Ki chunks

DISCLAIMER: This result is a bit unexpected. If I understand correctly, mmaping a file should use the page cache, but we see only anonymous memory used. This result will require more investigation that I haven't had time for yet.

![Memory Usage Over Time](https://github.com/user-attachments/assets/a3a7bb1b-92ef-4228-b1ad-69c10096038d)
![PSI Change Over Time](https://github.com/user-attachments/assets/5fcb14b6-e037-4b0b-a652-913f08ae0229)

## Dirty File - No Swap, No memory.high
Perform 128Mi of buffered write to a new file, and then continuously allocate anonymous memory on the heap in 200Ki chunks

![Memory Usage Over Time](https://github.com/user-attachments/assets/43798422-b06c-4144-991c-b2deeddcffac)
![PSI Change Over Time](https://github.com/user-attachments/assets/311393c3-eff5-4e2d-8f78-cfcd0b9dd48e)

# Summary of No memory.high, No Swap Test

In a setup without memory.high set, and without any swap space, the only scenario in which the PSI metrics warn us of an impending OOM is when there are pages in the page cache being written back to disk (as seen in the dirty file test case).

In the cases that only use anonymous memory, we see a small PSI spike at about the same time as the OOM occurs. It's not totally clear to me if this happens immediately before or as a result of the OOM, so that could/would require more research.

**Comment by ryansmick** [comment]
# Test Bench
OS Version: Ubuntu 24.04
CPU Arch: x86
Kernel Version: 6.8.0-1016-aws
Cgroup memory limit (memory.max): 128Mi
Cgroup memory.high: 64Mi
Cgroup Swap Space (memory.swap.max): None

## Anonymous - No Swap, 64Mi memory.high
Continuously allocate anonymous memory on the heap in 200Ki chunks

![Memory Usage Over Time](https://github.com/user-attachments/assets/ca0d3d34-cc3a-4520-b7e6-e824c454592b)
![Aggregate PSI Change Over Time](https://github.com/user-attachments/assets/1db95dc5-c12c-49f3-a5ea-97309f870253)
![PSI Change Over Time](https://github.com/user-attachments/assets/76b5ab4b-c131-490d-8b24-0c0c074ecaba)

## Clean File - No Swap, 64Mi memory.high
Perform buffered read of 128Mi file, and then continuously allocate anonymous memory on the heap in 200Ki chunks

![Memory Usage Over Time](https://github.com/user-attachments/assets/881cfc5d-c98a-45e2-8e6d-6fe5a4ed9d3b)
![Aggregate PSI Change Over Time](https://github.com/user-attachments/assets/b9f0450a-fe2c-4099-9c3e-7f44edd6e341)
![PSI Change Over Time](https://github.com/user-attachments/assets/8725482f-556c-4143-8940-9a9f220beaca)

## Dirty File - No Swap, 64Mi memory.high
Perform 128Mi of buffered write to a new file, and then continuously allocate anonymous memory on the heap in 200Ki chunks

![Memory Usage Over Time](https://github.com/user-attachments/assets/59d9f476-5b42-4492-89f3-c07b2b254f55)
![Aggregate PSI Change Over Time](https://github.com/user-attachments/assets/17de02a8-4646-45e7-84f1-1f2f897957eb)
![PSI Change Over Time](https://github.com/user-attachments/assets/d515d046-4ba2-4b46-b001-54a39637082b)

# Summary of 64Mi memory.high, No Swap Test

When the memory.high threshold is crossed, we see a linear increase in microseconds of stall time of the process. Based on my calculations, the percentage of stall time is about 99%, meaning that for every millisecond of real-world time, the CPU is only spending 10 microseconds on that process/task. The aggregated PSI metrics (avg10, avg60, avg300) seem to asymptotically approach this 99% value, but note that this is just an experimental result and I don't yet know how memory.high's throttling is implemented.

It's difficult to see from the graph, but it's also worth noting that there's about a 1.5 second gap between when the microsecond stall metrics (suffixed with `_total`) start incrementing and when the aggregated PSI metrics start incrementing.

**Comment by KBaichoo** [comment]
Thanks you @ryansmick for running these experiments, they look great! 

When you do the continuous allocation of heap, how are you achieving this? e.g.

Something like:
```
while {
  chunk = malloc(200kb)
  for page size in chunk:
    touch chunk setting 1 byte
}
```
?

Likewise for the 128Mi buffered write, how did you achieve this?

> Regarding the no memory.high experiements

It seems that the PSI doesn't give us much signal until too late in most cases.

The issue with the clean file pages might be related to using the file system page cache or some other dynamics.

> Regarding the memory.high experiements:

> Dirty file - no swap, 64MI memory high

I was initially surprised that this only hit 60MB and flat lined yet we see no throttling from PSI. This might be due to the system only lazily giving back the page memory though it might have already been written to the underlying file system.

More generally:

Users would need to tune or configure a memory.high in order to get some signal as the throttling and PSI signal works much better in these experiments. 

This does look promising, as the memory.high could be adjusted or some non-OS agent could take action. This could also trigger the process itself to take some action e.g. trigger Envoy overload.

For the synthetic program used I _think_ the performance degrades sharply due to the throttling kicking off direct reclaim by none of the pages can be freed since they're all anonymous process memory and there is no swap and all the program is trying to do at that point is allocate more memory. I'd expect more real world programs might more gradually degrade since they'll likely have a heap allocator and not constantly be going to the OS for large chunks of memory.

**Comment by ryansmick** [comment]
Hey @KBaichoo , apologies for the delay. Ended up taking some extended time off for the holidays, and getting back to this now.

> When you do the continuous allocation of heap, how are you achieving this?

I ended up writing a go program to do all of this. Let me add some code snippets below to illustrate how I achieved each part.

```golang
// Allocates 128Mi of anonymous memory by extending an array by 200Ki every tenth of a second.
func alloc_anon() {
	var a []byte
	for {
		a = append(a, make([]byte, BYTES_PER_ALLOCATION)...)
		time.Sleep(ALLOCATION_INTERVAL)
	}
}

// Allocates 128Mi of file-backed memory by reading a 128Mi file from disk, 200Ki every tenth of a second.
func alloc_clean_file() {
	f, err := os.Open("/tmp/clean_file")
	if err != nil {
		panic(err)
	}

	b := make([]byte, BYTES_PER_ALLOCATION)

	for {
		_, err := f.Read(b)
		if err != nil && !errors.Is(err, io.EOF) {
			fmt.Println(err)
			break
		}
		if err != nil {
			// end of file
			break
		}
		time.Sleep(ALLOCATION_INTERVAL)
	}
}

func alloc_clean_mmap_file() {
	reader, err := mmap.Open("/tmp/clean_file") // "golang.org/x/exp/mmap"
	if err != nil {
		panic(err)
	}
	defer reader.Close()

	// Read the contents of the file
	iterations := 0
	b := make([]byte, BYTES_PER_ALLOCATION)
	for {
		_, err := reader.ReadAt(b, int64(iterations*BYTES_PER_ALLOCATION))
		if err != nil {
			fmt.Println(err)
			break
		}
		iterations++
		time.Sleep(ALLOCATION_INTERVAL)
	}
}

// Allocates 128Mi of file-backed memory by writing 200Ki to a file on disk every tenth of a second.
func alloc_dirty_file(f *os.File) {
	for i := 0; i < MEMORY_LIMIT/BYTES_PER_ALLOCATION; i++ {
		randBytes := make([]byte, BYTES_PER_ALLOCATION)
		_, err := rand.Read(randBytes)
		if err != nil {
			panic(err)
		}
		_, err = f.Write(make([]byte, BYTES_PER_ALLOCATION))
		if err != nil {
			panic(err)
		}

		time.Sleep(ALLOCATION_INTERVAL)
	}
}
```

I'd have to strace to view the exact system calls being used under the hood, but haven't gotten to that yet. I'm also interested in doing this for the mmaped file as well, because those results were a bit unexpected based on my understanding of mmaped files.

> It seems that the PSI doesn't give us much signal until too late in most cases.

Looking at the memory.high and the swap space testing (which I'm going to post here in a second), I think my conclusion is a  bit different. We get pretty instantaneous signal from the `some_total` and `full_total` metrics. The main issue is that **we need something to cause throttling before we OOM** (otherwise you're right, as we see in the no swap no memory.high case, we see no PSI change before the OOM occurs). That *something* could be a memory.high value, or just the presence of swap space (as you'll see in this next message). So my conclusion is that we actually can get notified early enough to do something.

But my fear (and this has yet to be tested) is that once we take action in Overload Manager (like stopping all new connections, for example), the PSI will stop changing because we're no longer allocating new memory, and so our trigger will actually go back to an un-triggered state, allowing new connections, re-triggering, and so on. Just continuously flapping. Again, this is untested, so I'd like to look into this (as well as the strace on the tests above) once I have some time.

> I was initially surprised that this only hit 60MB and flat lined yet we see no throttling from PSI. This might be due to the system only lazily giving back the page memory though it might have already been written to the underlying file system.

Good eye, that is interesting. I looked back at the raw data and there actually was some PSI change. About 17k microseconds of stall time between when inactive_file first hit 60Mi and when it hit 0Mi, which is nearly the same as what we saw in the no memory.high case. It just so happens that in this case the PSI skyrockets afterwards because of the presence of memory.high, so we can't see it on the graph.

> More generally...

Totally agree here. PSI signal works much better if we have something to throttle before OOM, like memory.high or swap. I agree that the test bench is pretty aggressive in attempting to allocate memory and real world programs would probably see a less-sharp drop in performance. Also agree this looks promising for Overload Manager.

There are still a couple things I want to look into like the strace for the test bench and attempting to understand how this would behave if implemented in Overload Manager (w.r.t the flapping I mentioned above). So maybe next steps are for me to post the swap test bench results, dig into the strace a bit, and then if everything looks good from there we can discuss more specifics about how we'd want to utilize/implement this in Envoy/Overload Manager. Let me know what you think! Thanks!

**Comment by ryansmick** [comment]
# Test Bench
OS Version: Ubuntu 24.04
CPU Arch: x86
Kernel Version: 6.8.0-1016-aws
Cgroup memory limit (memory.max): 128Mi
Cgroup memory.high: None
Cgroup Swap Space (memory.swap.max): 256Mi

## Anonymous - 256Mi Swap, No memory.high
Continuously allocate anonymous memory on the heap in 200Ki chunks

![Memory Usage Over Time](https://github.com/user-attachments/assets/9e6ef9f8-2959-440d-bb81-de01da2eba42)
![Aggregate PSI Change Over Time](https://github.com/user-attachments/assets/93e9aa4d-7be0-485f-b1e8-5c2254bfd019)
![PSI Change Over Time](https://github.com/user-attachments/assets/8cbb0817-2cc0-4e7a-9a55-b3f265f89c2a)

## Clean File - 256Mi Swap, No memory.high
Perform buffered read of 128Mi file, and then continuously allocate anonymous memory on the heap in 200Ki chunks

![Memory Usage Over Time](https://github.com/user-attachments/assets/1918b671-a1dc-494f-be28-0ee27073e572)
![PSI Change Over Time](https://github.com/user-attachments/assets/caf9ee11-cd64-4b93-a909-b467de8e328f)
![Aggregate PSI Change Over Time](https://github.com/user-attachments/assets/eef5746d-c58b-4d8f-8936-83bfda97bae1)

## Dirty File - 256Mi Swap, No memory.high
Perform 128Mi of buffered write to a new file, and then continuously allocate anonymous memory on the heap in 200Ki chunks

![Memory Usage Over Time (3)](https://github.com/user-attachments/assets/e7645f95-6865-46da-8c30-23fb6106901c)
![PSI Change Over Time (3)](https://github.com/user-attachments/assets/6e803612-eaa8-44ce-91c6-0742f834585e)
![Aggregate PSI Change Over Time (3)](https://github.com/user-attachments/assets/da308f57-5acd-4e0d-b053-769cdd88110d)

# Summary of No memory.high, 256Mi Swap Test

When the memory.max threshold is crossed, we see usage of the swap space begin. As swap is used, the memory PSI is increasing. During periods of steep increase in the "total" PSI metrics, we also see the aggregate PSI metrics increase. Swap space seems to be used in a pattern resembling a staircase (some steep increases, followed by flat sections). I don't know what about the system causes this pattern, but we see PSI responding accordingly: "total" metrics are incrementing during the swap usage increases and remaining constant while swap use is constant, and the aggregate metrics exhibiting different behavior based on their lookback (the 10s average decreases during periods of constant swap usage while the 300s average is mostly monotonically increasing).

**Comment by KBaichoo** [comment]
Hey @ryansmick ,

Sorry for the delay, will parse the rest of the results later but wanted to post on the code snippets.

Thank you for adding the code snippets.

Regarding alloc_clean_file:

I think this would just allocate 200Ki since we'd be re-using the 200Ki buffer for subsequent reads? From the comment it seems like BYTES_PER_ALLOCATION = 200Ki and Allocation interval = 100ms.

Likewise with alloc_clean_mmap_file, seems like we'd be reusing the buffer.

For alloc_dirty_file it seems like we don't use the randBytes but instead use zeroed bytes when we write the file?

**Comment by ryansmick** [comment]
@KBaichoo You're right about the clean reads. Really good catch. I misunderstood the file Read function. Will rerun these tests next week and see how the results change.

Same for the dirty file. Looks like I'm not actually using randBytes. I _think_ the results will be the same, but I will rerun that next week too and update the results if they change

**Comment by wujiaqi** [comment]
I'm very interested in this conversation. We've been exploring using overload manager to prevent OOMKills with load shedding and deploy using Istio as well. However I have been running into problems with the existing fixed_heap resource monitor.

In my local experiments I've found that the pressure reported via the heap resource monitor and the memory allocation that docker sees is very disconnected and I assume this translates into our kubernetes deployment as well. I don't fully understand the details but I understand this is related to how tmalloc works and its relationship with the OS's view of memory allocated to the process. This makes the existing overload manager resource monitors on memory untenable for use in Istio deployments. 

However catching up on what I've read here, I'm also noticing that the proposed solution here depends on cgroups v2 which unfortunately our kubernetes deployments do not support yet. I would like for some sort of cgroup aware solution. I saw that there was a contribution recently to make the cpu monitor cgroup aware which is great. Having something available for memory would be great as well and would be even better if it were widely compatible, like with cgroups v1. We plan to get to cgroups v2 potentially this year but it may take some time.

**Comment by KBaichoo** [comment]
Hi @wujiaqi ,

Don't think you need to block on cgroupv2. I think you could create a cgroup aware memory monitor

> What if you instead created a cgroup based heap monitor that would dynamically adjust to the resources provides to the group -- by reading the value itself and providing some fixed percent of that allocated for envoy to react.

from https://github.com/envoyproxy/envoy/issues/36681#issuecomment-2433461714

Happy to provide reviews for both of these.

**Comment by wujiaqi** [comment]
@KBaichoo I opened a separate issue to discuss this PTAL thanks https://github.com/envoyproxy/envoy/issues/38718

**Comment by ryansmick** [comment]
@KBaichoo Just completed rerunning the no memory.high, no swap test cases for clean, clean-mmap, and dirty files after making the code adjustments you mentioned above. For the sake of brevity (since not much changed), I'll describe the changes and the test cases below:

# Clean File
For both of these test cases, I fixed a bug where I wasn't seeking into the file. The code snipped just reread the same bytes from the file over and over again on every iteration of the loop. But that is fixed for both of the below tests.

## Test Case 1: Allocate a new 200KiB buffer on each iteration of the loop
Very similar to the previous "Clean File - No Swap, No memory.high" test run but with a higher value on the flat left side. I imagine this is because Go is GCing the buffers that have now been dropped on previous iterations of the loop. Other than that, nothing really different from the previous test case. No PSI changes.

<img width="598" alt="Image" src="https://github.com/user-attachments/assets/3ae54b98-3c3f-4d41-9e98-f5115514e67b" />

## Test Case 2: Allocate a 128MiB buffer up front before starting the loop
Anonymous memory usage climbs until it gets OOM killed at 128MiB during the file read. Memory increase is mostly much more smooth than the previous test. No PSI changes.

<img width="599" alt="Image" src="https://github.com/user-attachments/assets/7ba0da56-6f91-4db9-ad83-3156fd3ccae9" />

# Clean Mmap File

## Test Case 1: Allocate a new 200KiB buffer on each iteration of the loop
Very similar to the previous "Clean Mmaped File - No Swap, No memory.high" test run but with a higher value on the flat left side. I imagine this is because Go is GCing the buffers that have now been dropped on previous iterations of the loop. Other than that, nothing really different from the previous test case. No PSI changes.

<img width="599" alt="Image" src="https://github.com/user-attachments/assets/fcf6b1ac-1559-4036-8677-f4fa673cd511" />

## Test Case 2: Allocate a 128MiB buffer up front before starting the loop
Anonymous memory usage climbs until it gets OOM killed at 128MiB during the file read. Interestingly it's a much more smooth increase than we've previously seen in any of the graphs. No PSI changes.

<img width="599" alt="Image" src="https://github.com/user-attachments/assets/10dd312a-e2a6-47d8-b372-7b4fe9902d8f" />

# Dirty File

## Test Case 1: Write randBytes to the output file
No change in memory allocation or PSI from previous "Dirty File - No Swap, No memory.high" test

<img width="599" alt="Image" src="https://github.com/user-attachments/assets/2ed5b795-6d3b-4c55-bd95-b0628241686a" />

---
<br>

With these results, if it sounds good to you I think I'll experiment a bit with how envoy reacts to increased load next (I imagine envoy will be using almost entirely anonymous memory, but worth a check), and I'll start thinking about some implementation details to propose here and discuss. Let me know if you have any thoughts on the above results or how to proceed next. Thanks!

**Comment by ryansmick** [comment]
@wujiaqi thanks for pushing a cgroup-aware monitor forward! I'm looking forward to experimenting with whatever you end up adding
